File: Advanced Programming Language/Multi Paradigm Statistics Solution/StatisticsCalculator.ml
let mean lst =
  let sum = List.fold_left (fun acc x -> acc + x) 0 lst in
  float_of_int sum /. float_of_int (List.length lst)

let median lst =
  let sorted = List.sort compare lst in
  let len = List.length sorted in
  if len mod 2 = 0 then
    float_of_int (List.nth sorted (len/2 - 1) + List.nth sorted (len/2)) /. 2.0
  else
    float_of_int (List.nth sorted (len/2))

let mode lst =
  let counts = List.fold_left (fun acc x ->
    if List.mem_assoc x acc then
      (x, List.assoc x acc + 1) :: (List.remove_assoc x acc)
    else (x, 1) :: acc) [] lst in
  let max_freq = List.fold_left (fun m (_, f) -> max m f) 0 counts in
  List.filter (fun (_, f) -> f = max_freq) counts |> List.map fst

(* Output Logic *)
let () =
  let data = [12; 5; 8; 12; 5; 10; 15] in
  print_endline "--- OCaml Functional Statistics ---";
  Printf.printf "Input Data: ";
  List.iter (Printf.printf "%d ") data;
  Printf.printf "\nMean: %.2f\n" (mean data);
  Printf.printf "Median: %.2f\n" (median data);
  Printf.printf "Mode(s): ";
  List.iter (Printf.printf "%d ") (mode data);
  print_newline ()

File: Advanced Programming Language/Multi Paradigm Statistics Solution/StatisticsCalculator.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

double get_mean(int arr[], int n) {
    double sum = 0;
    for (int i = 0; i < n; i++) sum += arr[i];
    return sum / n;
}

double get_median(int arr[], int n) {
    qsort(arr, n, sizeof(int), compare);
    if (n % 2 == 0)
        return (arr[n/2 - 1] + arr[n/2]) / 2.0;
    return arr[n/2];
}

void print_mode(int arr[], int n) {
    // Count frequencies in a sorted array
    int max_count = 0, current_count = 1;
    for (int i = 1; i < n; i++) {
        if (arr[i] == arr[i-1]) current_count++;
        else {
            if (current_count > max_count) max_count = current_count;
            current_count = 1;
        }
    }
    if (current_count > max_count) max_count = current_count;

    printf("Mode(s): ");
    current_count = 1;
    for (int i = 1; i < n; i++) {
        if (arr[i] == arr[i-1]) current_count++;
        else {
            if (current_count == max_count) printf("%d ", arr[i-1]);
            current_count = 1;
        }
    }
    if (current_count == max_count) printf("%d ", arr[n-1]);
    printf("\n");
}

int main() {
    int data[] = {12, 5, 8, 12, 5, 10, 15};
    int n = sizeof(data) / sizeof(data[0]);

    printf("--- C Procedural Statistics ---\n");
    printf("Input Data: ");
    for(int i=0; i<n; i++) printf("%d ", data[i]);
    
    printf("\nMean: %.2f\n", get_mean(data, n));
    printf("Median: %.2f\n", get_median(data, n));
    print_mode(data, n);
    
    return 0;
}
